### ISSUE TITLE
KL loss in the image-to-image translation

### ISSUE THREAD

**Issue by WERush** [mingyuliutw/UNIT#10]
Hi, mingyu. Good job.

I am confused about the [KL loss](https://github.com/mingyuliutw/UNIT/blob/master/src/trainers/cocogan_trainer.py#L32) in the image-to-image task. I found that this KL loss was different with [another one](https://github.com/mingyuliutw/UNIT/blob/master/src/trainers/cocogan_trainer_da.py#L45) in domain adaptation task. Can you explain it and where can I find the first KL loss in the paper? Thanks very much.

**Comment by mingyuliutw** [comment]
In the image translation experiments, we used a reduced VAE formulation. In the original VAE formulation, the encoder outputs a mean vector and a variance vector. Through the re-parameterization trick, the input to the generator is the element-wise sum of the mean vector and the element-wise multiplication of the variance vector and a noise vector sampled from N(0,I). In our reduced VAE form, we make the variance vector an all one vector. Hence, the input to the generator is then the element-wise sum of the mean vector and a noise vector sampled from N(0,I). We implemented this operation via a GaussianNoiseLayer, which can be found under src/trainers/common_net.py.

I hope this clear the doubts.

**Comment by gmkim90** [comment]
Thanks for providing the interesting paper and code.

I have other questions related to VAE loss.

Q1) What is the advantage of fixing variance of gaussian as 1 for VAE?
Q2) What is the advantage of choosing weights for KL:reconstruction = 1:1000? (either empirical or theoretical reason)
       If they are different from 1:1, then it cannot be garunteed to be upper bound of log-likelihood.

**Comment by mingyuliutw** [comment]
@gmkim90 Thanks for your interest on our paper.

For Q1, the prior distribution is a prior distribution. We fixed the variance to 1 for its simple form. We tested it with learnable variance. It works equally well. If you checked our SVHN2MNIST implementation, you would find an example of using learnable variance. 

For Q2, from my perspective, it does not make much sense to argue whether a particular set of $\lambda_1$ or $\lambda_2$ values violate the lower-bound assumption or not. Note that we can always absorb the weight factor into the hyperparameter in the data distribution (the scale parameter for Laplace distribution and the variance or squared scale parameter for normal distribution). For example, 
we can achieve the same effect by using 1:1 weight by making the scale parameter of the Laplace distribution to be 0.001.
